eth_tester/main.py

```diff
@@ -64,6 +69,30 @@ def get_default_fork_blocks():
     }
 
 
+def handle_auto_mining(func):
+    @functools.wraps(func)
+    def func_wrapper(self, *args, **kwargs):
+        if not self.auto_mine_transactions:
+            snapshot = self.take_snapshot()
+
+        transaction_hash = func(self, *args, **kwargs)
```

Comment:
I know this starts to make the function more awkward, but I think we should also revert if `func` to raise an exception too.

Maybe we should just have two main blocks, whether `auto_mine` is on, and just duplicate the `func()` call, like:

```py
+        if self.auto_mine_transactions:
+            transaction_hash = func(self, *args, **kwargs)
+            self.mine_block()
+        else:
+            snapshot = self.take_snapshot()
+            try:
+                transaction_hash = func(self, *args, **kwargs)
+                pending_transaction = self.get_transaction_by_hash(transaction_hash)
+                # Remove any pending transactions with the same nonce
+                self._pending_transactions = remove_matching_transaction_from_list(
+                    self._pending_transactions, pending_transaction)
+                self._pending_transactions.append(pending_transaction)
+            finally:
+                self.revert_to_snapshot(snapshot)
```

I find that easier to follow than the repeated "if auto-mine" tests.

Reply:
Ah yes that definitely is easier to read 😄